File: apps/android/app/src/main/java/com/termirust/mobile/controller/NativeControllerTerminal.kt
package com.termirust.mobile.controller

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

internal object NativeControllerTerminal {
    val loaded = runCatching { System.loadLibrary("termirust_mobile_ffi") }.isSuccess

    @JvmStatic external fun create(columns: Int, rows: Int, scrollbackRows: Int): Long
    @JvmStatic external fun process(handle: Long, input: ByteArray): ByteArray
    @JvmStatic external fun feed(handle: Long, input: ByteArray): Boolean
    @JvmStatic external fun resize(handle: Long, columns: Int, rows: Int): ByteArray
    @JvmStatic external fun snapshot(handle: Long): ByteArray
    @JvmStatic external fun destroy(handle: Long)
}

internal class NativeControllerTerminalSession private constructor(
    private var handle: Long,
) : AutoCloseable {
    fun process(bytes: ByteArray) = decode(NativeControllerTerminal.process(handle, bytes))

    fun feed(bytes: ByteArray) {
        check(handle != 0L) { "native terminal is closed" }
        check(NativeControllerTerminal.feed(handle, bytes)) { "native terminal rejected the frame" }
    }

    fun resize(viewport: TerminalViewport) = decode(
        NativeControllerTerminal.resize(handle, viewport.columns, viewport.rows),
    )

    fun snapshot() = decode(NativeControllerTerminal.snapshot(handle))

    override fun close() {
        if (handle == 0L) return
        NativeControllerTerminal.destroy(handle)
        handle = 0
    }

    private fun decode(bytes: ByteArray): BoundedTerminalSnapshot {
        check(handle != 0L) { "native terminal is closed" }
        return JSON.decodeFromString<NativeTerminalSnapshot>(bytes.decodeToString()).toModel()
    }

    companion object {
        private val JSON = Json { ignoreUnknownKeys = false }

        fun openOrNull(
            viewport: TerminalViewport,
            limits: TerminalLimits,
        ): NativeControllerTerminalSession? {
            if (!NativeControllerTerminal.loaded) return null
            val cellsPerRow = viewport.columns.coerceAtLeast(1)
            val cellBound = (limits.maxRetainedCells / cellsPerRow - viewport.rows).coerceAtLeast(0)
            val byteBound = (limits.maxModelBytes / (cellsPerRow * 64).coerceAtLeast(1) - viewport.rows)
                .coerceAtLeast(0)
            val scrollback = minOf(limits.maxScrollbackRows, cellBound, byteBound)
            return runCatching {
                NativeControllerTerminal.create(viewport.columns, viewport.rows, scrollback)
            }.getOrNull()?.takeIf { it != 0L }?.let(::NativeControllerTerminalSession)
        }
    }
}

@Serializable
private data class NativeTerminalSnapshot(
    @SerialName("schema_version") val schemaVersion: Int,
    val columns: Int,
    val rows: Int,
    val lines: List<String>,
    val cells: List<List<NativeTerminalCell>>,
    @SerialName("cursor_row") val cursorRow: Int,
    @SerialName("cursor_column") val cursorColumn: Int,
    @SerialName("cursor_visible") val cursorVisible: Boolean,
    @SerialName("application_cursor") val applicationCursor: Boolean,
    @SerialName("application_keypad") val applicationKeypad: Boolean,
    @SerialName("alternate_screen") val alternateScreen: Boolean,
    @SerialName("bracketed_paste") val bracketedPaste: Boolean,
    @SerialName("mouse_mode") val mouseMode: String,
    @SerialName("mouse_encoding") val mouseEncoding: String,
    @SerialName("scrollback_rows") val scrollbackRows: Int,
    @SerialName("retained_cells") val retainedCells: Int,
    @SerialName("accounted_bytes") val accountedBytes: Int,
) {
    fun toModel(): BoundedTerminalSnapshot {
        check(schemaVersion == 1) { "unsupported native terminal snapshot schema" }
        check(columns > 0 && rows > 0 && cells.size >= rows)
        val content = cells.map { row -> row.map(NativeTerminalCell::toModel) }
        check(content.all { it.size <= columns })
        val padded = content.map { row ->
            row + List(columns - row.size) { BoundedTerminalCell.blank() }
        }
        return BoundedTerminalSnapshot(
            lines = lines,
            cells = padded,
            contentCells = content,
            cursorRow = cursorRow,
            cursorColumn = cursorColumn,
            retainedCells = retainedCells,
            accountedBytes = accountedBytes,
            truncation = null,
            cursorVisible = cursorVisible,
            applicationCursor = applicationCursor,
            applicationKeypad = applicationKeypad,
            alternateScreen = alternateScreen,
            bracketedPaste = bracketedPaste,
            mouseMode = TerminalMouseMode.entries.first { it.wireName == mouseMode },
            mouseEncoding = TerminalMouseEncoding.entries.first { it.wireName == mouseEncoding },
            scrollbackRows = scrollbackRows,
        )
    }
}

@Serializable
private data class NativeTerminalCell(
    val text: String,
    val width: Int,
    val foreground: NativeTerminalColor,
    val background: NativeTerminalColor,
    val bold: Boolean,
    val dim: Boolean,
    val italic: Boolean,
    val underline: Boolean,
    val inverse: Boolean,
) {
    fun toModel() = BoundedTerminalCell(
        text = text,
        width = TerminalCellWidth.entries.first { it.columns == width },
        style = TerminalCellStyle(
            foreground = foreground.toModel(),
            background = background.toModel(),
            bold = bold,
            dim = dim,
            italic = italic,
            underline = underline,
            inverse = inverse,
        ),
    )
}

@Serializable
private data class NativeTerminalColor(
    val kind: String,
    val value: Int? = null,
    val red: Int? = null,
    val green: Int? = null,
    val blue: Int? = null,
) {
    fun toModel(): TerminalCellColor = when (kind) {
        "default" -> TerminalCellColor.Default
        "indexed" -> TerminalCellColor.Indexed(checkNotNull(value))
        "rgb" -> TerminalCellColor.Rgb(checkNotNull(red), checkNotNull(green), checkNotNull(blue))
        else -> error("unsupported native terminal color")
    }
}

private fun BoundedTerminalCell.isDefaultBlank() =
    text == " " && width == TerminalCellWidth.NARROW && style == TerminalCellStyle()

File: apps/ios/TermiRustMobile/Terminal/NativeControllerTerminal.swift
import Foundation
import TermiRustMobileCrypto

final class NativeControllerTerminalSession: @unchecked Sendable {
    private var handle: OpaquePointer?

    init?(
        viewport: TerminalViewportState,
        limits: TerminalLimits
    ) {
        let cellsPerRow = max(viewport.columns, 1)
        let cellBound = max(0, limits.maxRetainedCells / cellsPerRow - viewport.rows)
        let byteBound = max(0, limits.maxModelBytes / max(cellsPerRow * 64, 1) - viewport.rows)
        let scrollback = min(limits.maxScrollbackRows, cellBound, byteBound)
        handle = termirust_mobile_terminal_create(
            UInt16(clamping: viewport.columns),
            UInt16(clamping: viewport.rows),
            scrollback
        )
        if handle == nil { return nil }
    }

    deinit {
        termirust_mobile_terminal_destroy(handle)
    }

    func process(_ data: Data) throws -> BoundedTerminalSnapshot {
        guard let handle else { throw NativeTerminalError.closed }
        let result = data.withUnsafeBytes { bytes in
            termirust_mobile_terminal_process(
                handle,
                bytes.bindMemory(to: UInt8.self).baseAddress,
                data.count
            )
        }
        return try decode(result)
    }

    func feed(_ data: Data) throws {
        guard let handle else { throw NativeTerminalError.closed }
        let accepted = data.withUnsafeBytes { bytes in
            termirust_mobile_terminal_feed(
                handle,
                bytes.bindMemory(to: UInt8.self).baseAddress,
                data.count
            )
        }
        guard accepted else {
            throw NativeTerminalError.operationFailed("Native terminal rejected the frame.")
        }
    }

    func resize(_ viewport: TerminalViewportState) throws -> BoundedTerminalSnapshot {
        guard let handle else { throw NativeTerminalError.closed }
        return try decode(termirust_mobile_terminal_resize(
            handle,
            UInt16(clamping: viewport.columns),
            UInt16(clamping: viewport.rows)
        ))
    }

    func snapshot() throws -> BoundedTerminalSnapshot {
        guard let handle else { throw NativeTerminalError.closed }
        return try decode(termirust_mobile_terminal_snapshot(handle))
    }

    private func decode(_ result: TermiRustMobileResult) throws -> BoundedTerminalSnapshot {
        defer { termirust_mobile_free_result(result) }
        guard result.ok else {
            throw NativeTerminalError.operationFailed(String(
                decoding: data(from: result.error),
                as: UTF8.self
            ))
        }
        return try JSONDecoder().decode(
            NativeTerminalSnapshot.self,
            from: data(from: result.data)
        ).model()
    }

    private func data(from buffer: TermiRustMobileByteBuffer) -> Data {
        guard let pointer = buffer.ptr, buffer.len > 0 else { return Data() }
        return Data(bytes: pointer, count: buffer.len)
    }
}

private enum NativeTerminalError: Error {
    case closed
    case operationFailed(String)
}

private struct NativeTerminalSnapshot: Decodable {
    let schemaVersion: Int
    let columns: Int
    let rows: Int
    let lines: [String]
    let cells: [[NativeTerminalCell]]
    let cursorRow: Int
    let cursorColumn: Int
    let cursorVisible: Bool
    let applicationCursor: Bool
    let applicationKeypad: Bool
    let alternateScreen: Bool
    let bracketedPaste: Bool
    let mouseMode: String
    let mouseEncoding: String
    let scrollbackRows: Int
    let retainedCells: Int
    let accountedBytes: Int

    enum CodingKeys: String, CodingKey {
        case columns, rows, lines, cells
        case schemaVersion = "schema_version"
        case cursorRow = "cursor_row"
        case cursorColumn = "cursor_column"
        case cursorVisible = "cursor_visible"
        case applicationCursor = "application_cursor"
        case applicationKeypad = "application_keypad"
        case alternateScreen = "alternate_screen"
        case bracketedPaste = "bracketed_paste"
        case mouseMode = "mouse_mode"
        case mouseEncoding = "mouse_encoding"
        case scrollbackRows = "scrollback_rows"
        case retainedCells = "retained_cells"
        case accountedBytes = "accounted_bytes"
    }

    func model() throws -> BoundedTerminalSnapshot {
        guard schemaVersion == 1,
              columns > 0,
              rows > 0,
              cells.count >= rows,
              let mouseMode = TerminalMouseMode(rawValue: mouseMode),
              let mouseEncoding = TerminalMouseEncoding(rawValue: mouseEncoding) else {
            throw NativeTerminalError.operationFailed("Unsupported native terminal snapshot.")
        }
        let content = try cells.map { try $0.map { try $0.model() } }
        guard content.allSatisfy({ $0.count <= columns }) else {
            throw NativeTerminalError.operationFailed("Invalid native terminal row width.")
        }
        let padded = content.map { row in
            row + Array(repeating: .blank(), count: columns - row.count)
        }
        return BoundedTerminalSnapshot(
            lines: lines,
            cells: padded,
            contentCells: content,
            cursorRow: cursorRow,
            cursorColumn: cursorColumn,
            retainedCells: retainedCells,
            accountedBytes: accountedBytes,
            truncation: nil,
            cursorVisible: cursorVisible,
            applicationCursor: applicationCursor,
            applicationKeypad: applicationKeypad,
            alternateScreen: alternateScreen,
            bracketedPaste: bracketedPaste,
            mouseMode: mouseMode,
            mouseEncoding: mouseEncoding,
            scrollbackRows: scrollbackRows
        )
    }
}

private struct NativeTerminalCell: Decodable {
    let text: String
    let width: Int
    let foreground: NativeTerminalColor
    let background: NativeTerminalColor
    let bold: Bool
    let dim: Bool
    let italic: Bool
    let underline: Bool
    let inverse: Bool

    func model() throws -> BoundedTerminalCell {
        guard let width = TerminalCellWidth(rawValue: width) else {
            throw NativeTerminalError.operationFailed("Unsupported native terminal cell width.")
        }
        return BoundedTerminalCell(
            text: text,
            width: width,
            style: TerminalCellStyle(
                foreground: try foreground.model(),
                background: try background.model(),
                bold: bold,
                dim: dim,
                italic: italic,
                underline: underline,
                inverse: inverse
            )
        )
    }
}

private struct NativeTerminalColor: Decodable {
    let kind: String
    let value: UInt8?
    let red: UInt8?
    let green: UInt8?
    let blue: UInt8?

    func model() throws -> TerminalCellColor {
        switch kind {
        case "default": return .default
        case "indexed":
            guard let value else { break }
            return .indexed(value)
        case "rgb":
            guard let red, let green, let blue else { break }
            return .rgb(red: red, green: green, blue: blue)
        default: break
        }
        throw NativeTerminalError.operationFailed("Unsupported native terminal color.")
    }
}

private extension BoundedTerminalCell {
    var isDefaultBlank: Bool {
        text == " " && width == .narrow && style == TerminalCellStyle()
    }
}
